### PR TITLE
Add Redis Serialization Protocol support

### PR DESCRIPTION
**Pull request by mertyildiran**
This PR adds the [Redis Serialization Protocol](https://redis.io/topics/protocol) support.

Screenshot from 2021-09-22 01-43-32

Screenshot from 2021-09-22 01-43-36

Screenshot from 2021-09-22 01-43-47

### CODE REVIEW

File: tap/extensions/redis/read.go [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+	} else if strings.HasPrefix(msg, noscriptPrefix) {
+		return nil, newNoScriptError(msg)
+	}
+	return nil, newDataError(msg)
+}
+
+func (p *RedisProtocol) parseTargetHostAndSlot(clusterRedirectResponse string) (string, int, int) {
+	arr := strings.Split(clusterRedirectResponse, " ")
+	host, port := p.extractParts(arr[2])
+	slot, _ := strconv.Atoi(arr[1])
```
Comment (nimrod-up9) [inline tap/extensions/redis/read.go:457]: Check error.

File: tap/extensions/redis/helpers.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	return json.Marshal(h.Data)
+}
+
+type RedisWrapper struct {
+	Method  string      `json:"method"`
+	Url     string      `json:"url"`
+	Details interface{} `json:"details"`
+}
+
+func representGeneric(generic map[string]interface{}) (representation []interface{}) {
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/extensions/redis/helpers.go]: @nimrod-up9 done with https://github.com/up9inc/mizu/commit/3c070b4bc0c821a353ef063343eb5b6afee64162

File: tap/extensions/redis/read.go [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+	clusterSlaves           = "slaves"
+	clusterFailOver         = "failover"
+	clusterSlots            = "slots"
+	pubSubChannels          = "channels"
+	pubSubNumSub            = "numsub"
+	pubSubNumPat            = "numpat"
+)
+
+//IntToByteArr convert int to byte array
+func IntToByteArr(a int) []byte {
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/extensions/redis/read.go]: @nimrod-up9 done with https://github.com/up9inc/mizu/commit/6ac289a5fe1659b0ef699c603dc57f541bd104cf

File: tap/extensions/redis/main.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	extension.Protocol = &protocol
+	extension.MatcherMap = reqResMatcher.openMessagesMap
+}
+
+func (d dissecting) Ping() {
+	log.Printf("pong %s\n", protocol.Name)
+}
+
+func (d dissecting) Dissect(b *bufio.Reader, isClient bool, tcpID *api.TcpID, counterPair *api.CounterPair, superTimer *api.SuperTimer, superIdentifier *api.SuperIdentifier, emitter api.Emitter, options *api.TrafficFilteringOptions) error {
+	is := &RedisInputStream{
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/extensions/redis/main.go:41]: @nimrod-up9 done with https://github.com/up9inc/mizu/commit/2cee8070b1c31becad5599f442dbbba8a1114ea3

File: tap/extensions/redis/read.go (hunk 1 above)
Comment (mertyildiran) [reply tap/extensions/redis/read.go]: @nimrod-up9 done with https://github.com/up9inc/mizu/commit/a97b41949fd4608e35460136e26f352e448869e6

File: tap/extensions/redis/read.go [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+			c := p.is.Buf[p.is.count]
+			p.is.count++
+			if c != '\n' {
+				return nil, newConnectError("Unexpected character!")
+			}
+			break
+		} else {
+			line = append(line, b)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/extensions/redis/read.go:401]: @nimrod-up9 I've moved this file to here from https://github.com/piaohao/godis/blob/master/protocol.go and modified a little bit according to our need.

I didn't change the structure much. The `RedisProtocol` was a struct that's wrapping `redisOutputStream` and `redisInputStream` originally:

```go
type protocol struct {
	os *redisOutputStream
	is *redisInputStream
}
```

that's why it's like that.

File: tap/extensions/redis/helpers.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	details, _ := json.Marshal([]map[string]string{
```
[the remark this answers is not in the document]
Comment (mertyildiran) [reply tap/extensions/redis/helpers.go:28]: @nimrod-up9 this cannot fail on runtime and at this moment, there is no way of propagating the error back. Also it's like that in many places, inside other extensions right now:

- https://github.com/up9inc/mizu/blob/develop/tap/extensions/amqp/helpers.go
- https://github.com/up9inc/mizu/blob/develop/tap/extensions/http/main.go#L234-L377
- https://github.com/up9inc/mizu/blob/develop/tap/extensions/kafka/helpers.go

**Review by nimrod-up9 — APPROVED** [review]

**Review by nimrod-up9 — APPROVED** [review]

**Pull request merged**